### PR TITLE
xspress3 device based on hxntools

### PR DESCRIPTION
**Pull request by mrakitin**
Work by @tacaswell and @lcdesilva.

### CODE REVIEW

**Comment by mrakitin** [comment]
Tom, this is working now. I didn't use the generate_datums method from SRX, but maybe we don't need it here. Do you mind reviewing it before Thursday? Thanks.

File: startup/40-xspress3.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+            xs_det.stage()
+            # TODO: set the "XF:07BMB-ES{Xsp:1}:HDF5:FileWriteMode" PV to "Stream" (option 2)
+            xs_det.settings.num_images.put(self.num_points[xs_det.name])
+            xs_det.hdf5.num_capture.put(self.num_points[xs_det.name])
+
+        # self.det.external_trig.put(True)  # This is questionable, if we need it or not.
+        # Next line should take care of everything.
+        # xspress3 CSS:
+        for xs_det in self.xs_dets:
+            xs_det.settings.trigger_mode.put(3)  # "Acquisition Controls and Status" (top left pane) -->
```
Comment (tacaswell) [inline startup/40-xspress3.py]: We have a slight preference for using the enum strings in the code.  It is a bit less stable API wise, but much more readable.

File: startup/40-xspress3.py [hunk 2]
```diff
@@ -143,12 +148,24 @@ def complete(self, *args, **kwargs):
...
+
+        for datum_id, ts in zip(self._datum_ids, di_timestamps):
             data = {self.name: datum_id}
+            # TODO: fix the lost precision as pymongo complained about np.float128.
+            ts = float(ts)
             yield {'data': data,
-                   'timestamps': {key: now for key in data},
-                   'time': now,  # TODO: use the proper timestams from the mono start and stop times
+                   'timestamps': {key: ts for key in data},
+                   'time': ts,  # TODO: use the proper timestams from the mono start and stop times
```
Comment (mrakitin) [inline startup/40-xspress3.py]:
```suggestion
                   'time': ts,  # TODO: use the proper timestamps from the mono start and stop times
```

File: startup/40-xspress3.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+            self._asset_docs_cache.append(('datum', datum))
+            self._datum_ids.append(datum_id)
+
+        return NullStatus()
+
+    def collect(self):
+        # TODO: try to separate it from the xspress3 class
+        collected_frames = self.settings.array_counter.get()
+        dpb_sec_values = np.array(dpb_sec.get(), dtype='float128')[:collected_frames * 2: 2]
+        dpb_nsec_values = np.array(dpb_nsec.get(), dtype='float128')[:collected_frames * 2: 2]
```
Comment (mrakitin) [inline startup/40-xspress3.py]: Discussed it with @danielballan and @tacaswell, agreed to use these values as integers, and store them in `data` as separate columns.

The timestamps stores in databroker would be the timestamp of the reading of these waveforms from the IOC. Need to use `.get_with_metadata(...)`:
```suggestion
        dpb_nsec_values = np.array(dpb_nsec.get_with_metadata(), dtype='float128')[:collected_frames * 2: 2]
```

**Comment by mrakitin** [comment]
Based on the conversation with @klauer in the Nikea slack, this can be considered to have a consistent return from `.get()`:
```py
In [1]: class DerivedEpicsSignal(EpicsSignal):
   ...:     def __init__(self, *args, nelm_signal, **kwargs):
   ...:         super().__init__(*args, **kwargs)
   ...:         self._nelm_signal = nelm_signal
   ...:
   ...:     def get(self, *args, **kwargs):
   ...:         nelm_count = int(self._nelm_signal.get())
   ...:         return super().get(*args, count=nelm_count, **kwargs)
   ...:

In [2]: tmp = DerivedEpicsSignal('XF:07BMB-CT{Enc02:DI}T:nsec_Bin_', nelm_signal=dpb_nsec_nelm, name='tmp')

In [3]: len(tmp.get())
Out[3]: 16384
```
This way we don't have to pass `count=...` explicitly to each `.get()` call. However, it imposes coupling of signals... Not sure what is better. Maybe it should go to the parent `Device`-based object.

File: startup/40-xspress3.py (hunk 1 above)
Comment (mrakitin) [reply startup/40-xspress3.py]: We implemented this. Thank you Tom.

File: startup/40-xspress3.py (hunk 3 above)
Comment (mrakitin) [reply startup/40-xspress3.py]: This is still on our plate. Right now we are fine with losing the precision when summing seconds and nanoseconds (I think it's the current approach with pizza-boxes, when the timestamp is constricted from the seconds and nanoseconds from the resulted data file.)

**Comment by mrakitin** [comment]
Here is a better version of the code snippet, which allows to pass `count` explicitly to that derived signal:
```py
In [1]: class DerivedEpicsSignal(EpicsSignal):
   ...:     def __init__(self, *args, nelm_signal, **kwargs):
   ...:         super().__init__(*args, **kwargs)
   ...:         self._nelm_signal = nelm_signal
   ...:
   ...:     def get(self, *args, **kwargs):
   ...:         nelm_count = int(self._nelm_signal.get())
   ...:         count = kwargs.get('count', nelm_count)
   ...:         kwargs.update({'count': count})
   ...:         return super().get(*args, **kwargs)
   ...:

In [2]: tmp = DerivedEpicsSignal('XF:07BMB-CT{Enc02:DI}T:nsec_Bin_', nelm_signal=dpb_nsec_nelm, name='tmp')

In [3]: len(tmp.get())
Out[3]: 16384

In [4]: len(tmp.get(count=100))
Out[4]: 100
```

**Pull request merged**